File: lola/lola-sdks/typescript/src/context.ts
/**
 * Runtime context overrides: temporarily change chain, RPC URL, budget
 * limits, or mode for the duration of a callback, without mutating
 * global state.
 *
 *   await lola.withOverrides({ chain: "polygon", budgetMaxUsd: 5.0 }, async () => {
 *     const result = await executeContract(...);
 *   });
 *
 * Backed by Node's AsyncLocalStorage, so overrides correctly scope to the
 * async call tree of the callback — including across awaited promises —
 * without leaking into unrelated concurrent requests. In environments
 * without AsyncLocalStorage (older runtimes, some bundlers), this module
 * degrades to a simple module-level stack, which is still correct for
 * single-threaded synchronous nesting but cannot guarantee isolation
 * across concurrently-running async operations — see the README's
 * "Browser support" section.
 */

export interface Overrides {
  chain?: string;
  rpcUrl?: string;
  mode?: string;
  budgetMaxGas?: number;
  budgetMaxUsd?: number;
  hitlTimeoutSeconds?: number;
}

function mergeOverrides(base: Overrides, next: Overrides): Overrides {
  const merged: Overrides = { ...base };
  for (const [key, value] of Object.entries(next)) {
    if (value !== undefined) {
      (merged as Record<string, unknown>)[key] = value;
    }
  }
  return merged;
}

/** Shapes Overrides for inclusion in an RPC call's params, matching the
 * shape lola-core's config.Overrides expects on the wire. */
export function overridesToRpcParams(o: Overrides): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (o.chain !== undefined) out.chain = o.chain;
  if (o.rpcUrl !== undefined) out.rpc_url = o.rpcUrl;
  if (o.mode !== undefined) out.mode = o.mode;
  if (o.budgetMaxGas !== undefined) out.budget_max_gas = o.budgetMaxGas;
  if (o.budgetMaxUsd !== undefined) out.budget_max_usd = o.budgetMaxUsd;
  if (o.hitlTimeoutSeconds !== undefined) out.hitl_timeout_seconds = o.hitlTimeoutSeconds;
  return out;
}

// -- storage backend selection ---------------------------------------------

interface OverrideStorage {
  getStore(): Overrides | undefined;
  run<T>(value: Overrides, fn: () => T): T;
}

class AsyncLocalStorageBackend implements OverrideStorage {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private als: any;

  constructor() {
    // Loaded lazily/conditionally so bundlers targeting browsers (which
    // don't have node:async_hooks) don't choke on this import — see
    // FallbackStackBackend below for that environment.
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { AsyncLocalStorage } = require("async_hooks");
    this.als = new AsyncLocalStorage();
  }

  getStore(): Overrides | undefined {
    return this.als.getStore();
  }

  run<T>(value: Overrides, fn: () => T): T {
    return this.als.run(value, fn);
  }
}

class FallbackStackBackend implements OverrideStorage {
  private stack: Overrides[] = [];

  getStore(): Overrides | undefined {
    return this.stack[this.stack.length - 1];
  }

  run<T>(value: Overrides, fn: () => T): T {
    this.stack.push(value);
    try {
      return fn();
    } finally {
      this.stack.pop();
    }
  }
}

function createBackend(): OverrideStorage {
  try {
    return new AsyncLocalStorageBackend();
  } catch {
    return new FallbackStackBackend();
  }
}

const backend = createBackend();

/** Returns the currently active overrides (merged across nested calls),
 * or an empty object if none are set. */
export function getCurrentOverrides(): Overrides {
  return backend.getStore() ?? {};
}

/**
 * Runs `fn` (sync or async) with `overrides` applied for its entire
 * execution, composing with any enclosing `withOverrides` call (the
 * innermost override wins for any field both set).
 */
export function withOverrides<T>(overrides: Overrides, fn: () => T): T {
  const merged = mergeOverrides(getCurrentOverrides(), overrides);
  return backend.run(merged, fn);
}

File: lola/lola-sdks/python/lola_os/context.py
"""Runtime context overrides: temporarily change chain, RPC URL, budget
limits, or mode for the duration of a `with` block, without mutating
global state or requiring a config file edit.

    with lola.override(chain="polygon", budget_max_usd=5.0):
        result = execute_contract(...)

Overrides are stored in a contextvar so they compose correctly with
threads/async tasks and don't leak between unrelated calls.
"""

from __future__ import annotations

import contextvars
from contextlib import contextmanager
from dataclasses import dataclass, replace
from typing import Optional

_current_overrides: contextvars.ContextVar[Optional["Overrides"]] = contextvars.ContextVar(
    "lola_overrides", default=None
)


@dataclass(frozen=True)
class Overrides:
    chain: Optional[str] = None
    rpc_url: Optional[str] = None
    mode: Optional[str] = None
    budget_max_gas: Optional[float] = None
    budget_max_usd: Optional[float] = None
    hitl_timeout_seconds: Optional[int] = None

    def merged_with(self, other: "Overrides") -> "Overrides":
        """Returns a new Overrides with `other`'s non-None fields taking
        precedence over self's. Used to compose decorator-level overrides
        with an enclosing `with lola.override(...)` block."""
        updates = {k: v for k, v in other.__dict__.items() if v is not None}
        return replace(self, **updates)

    def to_rpc_params(self) -> dict:
        """Shapes this Overrides for inclusion in an RPC call's params
        under the `_context_overrides` key, matching the shape lola-core's
        config.Overrides expects."""
        out = {}
        if self.chain is not None:
            out["chain"] = self.chain
        if self.rpc_url is not None:
            out["rpc_url"] = self.rpc_url
        if self.mode is not None:
            out["mode"] = self.mode
        if self.budget_max_gas is not None:
            out["budget_max_gas"] = self.budget_max_gas
        if self.budget_max_usd is not None:
            out["budget_max_usd"] = self.budget_max_usd
        if self.hitl_timeout_seconds is not None:
            out["hitl_timeout_seconds"] = self.hitl_timeout_seconds
        return out


def get_current_overrides() -> Overrides:
    return _current_overrides.get() or Overrides()


@contextmanager
def override(
    chain: Optional[str] = None,
    rpc_url: Optional[str] = None,
    mode: Optional[str] = None,
    budget_max_gas: Optional[float] = None,
    budget_max_usd: Optional[float] = None,
    hitl_timeout_seconds: Optional[int] = None,
):
    """Context manager that scopes config overrides to the enclosed block.

    Example:

        with lola.override(rpc_url="https://backup.alchemy.io", chain="ethereum"):
            balance = get_balance("ethereum", "0x...")
    """
    new_overrides = Overrides(
        chain=chain,
        rpc_url=rpc_url,
        mode=mode,
        budget_max_gas=budget_max_gas,
        budget_max_usd=budget_max_usd,
        hitl_timeout_seconds=hitl_timeout_seconds,
    )
    merged = get_current_overrides().merged_with(new_overrides)
    token = _current_overrides.set(merged)
    try:
        yield merged
    finally:
        _current_overrides.reset(token)
